File: day10/Factory-2.py
from z3 import *
import re

INPUT_FILE = 'input.txt'

def factory_problem():
    with open(INPUT_FILE, 'r') as f:
        lines = f.readlines()
        solutions = []
        for line in lines:
            #print("Processing line:", line.strip())
            solutions.append(solve(line))
        #print(solutions)
        print("Solution:", sum(solutions))



def solve(line):
    (_, btns, jolts) = readMachineSpec(line)
    s = Optimize()
    num_buttons = len(btns)
    num_jolts = len(jolts)
    # Create Z3 integer variables for button presses
    button_vars = [Int(f'btn_{i}') for i in range(num_buttons)]
    # Add constraints for button presses
    for i in range(num_buttons):
        s.add(button_vars[i] >= 0)
    # Add constraints for joltages
    for i in range(num_jolts):
        s.add(jolts[i] == Sum([button_vars[b] for b in range(num_buttons) if i in btns[b]]))
    s.minimize(Sum(button_vars))

    if s.check() == sat:
        model = s.model()
        button_presses = [model[button_vars[i]].as_long() for i in range(num_buttons)]
        print(button_presses)
        return sum(button_presses)

    else:
        return None



# Vibede nedenstående med basis i min haskell-løsning
def split_commas(s):
    return [x for x in s.split(",") if x.strip()]

def inside_braces(open_c, close_c, s):
    """
    Extract substring inside matching open_c ... close_c,
    starting from the first occurrence of open_c.
    Returns (extracted_string, rest_after_closing).
    """
    start = s.find(open_c)
    if start == -1:
        return "", s
    depth = 0
    for i in range(start, len(s)):
        if s[i] == open_c:
            if depth == 0:
                content_start = i + 1
            depth += 1
        elif s[i] == close_c:
            depth -= 1
            if depth == 0:
                return s[content_start:i], s[i+1:]
    return "", s  # no matching close brace found


def readMachineSpec(s):
    # ------------------------------------------------------------
    # 1. Extract lights: a bracketed list of '.' and '#'
    # ------------------------------------------------------------
    # Equivalent to Haskell `lights`
    m = re.match(r"\s*\[([.#]*)\]", s)
    if not m:
        raise ValueError("Invalid or missing lights section `[ ... ]`")

    lights_str = m.group(1)
    rest = s[m.end():]

    # ------------------------------------------------------------
    # 2. Extract button lists: a sequence of `(x,y,...)`
    #    stopping before the `{...}` joltage section
    # ------------------------------------------------------------
    # Equivalent to Haskell `buttons`
    buttons = []
    while True:
        rest = rest.lstrip()
        if rest.startswith("{"):
            break  # end of button list

        if rest.startswith("("):
            nums_str, rest = inside_braces("(", ")", rest)
            nums = [int(x) for x in split_commas(nums_str)]
            buttons.append(nums)
        else:
            # skip non-brace characters (e.g., spaces)
            rest = rest[1:]

    # ------------------------------------------------------------
    # 3. Extract joltage list from `{...}`
    # ------------------------------------------------------------
    # Equivalent to Haskell `joltages`
    jolts_str, rest2 = inside_braces("{", "}", rest)
    joltages = [int(x) for x in split_commas(jolts_str)]

    return (lights_str, buttons, joltages)


if __name__ == "__main__":
    factory_problem()

File: day10/Factory-2.hs
import qualified Data.Map as Map
import Debug.Trace (trace)

----------------------------------------------
-- Some helper functions
split :: Char -> String -> [String]
split c = splitOnC []
  where
    splitOnC acc [] = [reverse acc]
    splitOnC acc (x : xs)
      | x == c = reverse acc : splitOnC [] xs
      | otherwise = splitOnC (x : acc) xs

insideBraces :: Char -> Char -> String -> (String, String)
insideBraces open close (o : s)
  | o == open = go s []
  where
    go (x : xs) acc
      | x == close = (reverse acc, xs)
      | x == open = go xs acc
      | otherwise = go xs (x : acc)

----------------------------------------------
-- Actual solution
----------------------------------------------

type MachineSpec = (String, [[Int]], [Int])

-- Read input
readMachineSpec :: String -> ([[Int]], [Int])
readMachineSpec = toMatrix . joltages . buttons . lights
  where
    lights :: String -> (String, String)
    lights (']' : rest) = ("", rest)
    -- lights ('[':xs) = lights xs
    lights (x : xs)
      | x `elem` ['.', '#'] = let (ls, rest) = lights xs in (x : ls, rest)
      | otherwise = lights xs

    buttons :: (String, String) -> (String, [[Int]], String)
    buttons (ls, ' ' : xs) = buttons (ls, xs) -- ignore spaces
    buttons (ls, rest@('{' : _)) = (ls, [], rest)
    buttons (ls, xs) =
      let (btn, rest) = button xs
          (nls, btns, finalRest) = buttons (ls, rest)
       in (nls, btn : btns, finalRest)
      where
        button :: String -> ([Int], String)
        button s =
          let (numsStr, rest) = insideBraces '(' ')' s
           in (map read $ split ',' numsStr, rest)

    joltages :: (String, [[Int]], String) -> MachineSpec
    joltages (ls, btns, rest) =
      let (numsStr, _) = insideBraces '{' '}' rest
       in (ls, btns, map read $ split ',' numsStr)

    toMatrix :: MachineSpec -> ([[Int]], [Int])
    toMatrix (ls, btns, js) =
      let m = length ls
          btnM = map (\b -> map (\i -> if i `elem` b then 1 else 0) [0 .. m - 1]) btns
       in (btnM, js)

-- Actual actual solution
fewestPresses :: ([[Int]], [Int]) -> Int
fewestPresses (btns, jolts) = trace "Entered function fewestPresses" fst $ tryAll jolts Map.empty
  where
    tryAll :: [Int] -> Map.Map [Int] Int -> (Int, Map.Map [Int] Int)
    tryAll js memo
      | Map.member js memo = (memo Map.! js, memo)
      | any (< 0) js = trace ("invalid found: " ++ show js) (maxBound :: Int, Map.insert js (maxBound :: Int) memo)
      | all (== 0) js = trace ("found solution: " ++ show js) (0, Map.insert js 0 memo)
      | otherwise =
          -- trace ("trying from state: " ++ show js) $
          foldl
            ( \(minPresses, curMemo) btn ->
                let (presses, newMemo) = tryAll (toggle js 0 btn) curMemo
                 in (min minPresses (presses + 1), newMemo)
            )
            (maxBound :: Int, memo)
            btns

    toggle :: [Int] -> Int -> [Int] -> [Int]
    toggle js _ [] = js
    toggle (l : ls) index (b : bs)
      | b == 1 =
          let nl = l - 1
           in nl : toggle ls (index + 1) bs
      | otherwise =
          l : toggle ls (index + 1) bs

main :: IO ()
main = do
  contents <- readFile "example.txt"
  let machines = lines contents
      matrices = map readMachineSpec machines
      result = map fewestPresses matrices
  print result

-- print result
